cirq-ft/cirq_ft/algos/select_swap_qroam.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2023 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import List, Optional, Sequence, Tuple
+
+import cirq
+import numpy as np
+from cirq._compat import cached_property
+from cirq_ft import infra
+from cirq_ft.algos import qrom, swap_network
+from numpy.typing import NDArray
+
+
+def find_optimal_log_block_size(iteration_length: int, target_bitsize: int) -> int:
+    """Find optimal block size, which is a power of 2, for SelectSwapQROM.
+
+    This functions returns the optimal `k` s.t.
+        * k is in an integer and k >= 0.
+        * iteration_length/2^k + target_bitsize*(2^k - 1) is minimized.
+    The corresponding block size for SelectSwapQROM would be 2^k.
+    """
+    k = 0.5 * np.log2(iteration_length / target_bitsize)
+    if k < 0:
+        return 1
+
+    def value(kk: List[int]):
+        return iteration_length / np.power(2, kk) + target_bitsize * (np.power(2, kk) - 1)
+
+    k_int = [np.floor(k), np.ceil(k)]  # restrict optimal k to integers
+    return int(k_int[np.argmin(value(k_int))])  # obtain optimal k
+
+
+@cirq.value_equality()
+class SelectSwapQROM(infra.GateWithRegisters):
```

Comment:
Should cirq-ft/cirq_ft/algos/select_swap_qroam.py be called cirq-ft/cirq_ft/algos/select_swap_qrom.py?

Reply:
done